### PR TITLE
Local handles initial drop

### PR DESCRIPTION
**Pull request by kumpera**
This is the local handles API I think we should be using. It addresses
issues from the current implementation and actually make it work.

### CODE REVIEW

**Comment by kumpera** [comment]
@lambdageek @ludovic-henry @vargaz

File: mono/metadata/handle.h [hunk 1]
```diff
@@ -16,183 +18,231 @@
...
+
+This is why we evaluate index and value before any call to MONO_HANDLE_RAW or other functions that deal with naked objects.
+*/
+#define MONO_HANDLE_SETRAW(HANDLE, FIELD, VALUE) do {	\
+		MonoObject *__val = (MonoObject*)(VALUE);	\
+		MONO_OBJECT_SETREF (MONO_HANDLE_RAW (HANDLE), FIELD, __val);	\
+	} while (0);
+
+#define MONO_HANDLE_SET(HANDLE, FIELD, VAL_HANDLE) do {	\
+		MONO_OBJECT_SETREF (MONO_HANDLE_RAW (HANDLE), FIELD, MONO_HANDLE_RAW (VAL_HANDLE));	\
```
Comment (lambdageek) [inline mono/metadata/handle.h]: Isn't this exactly what your comment says we shouldn't have? we'll have the raw ptr from `HANDLE` on the stack while we evaluate `VAL_HADNLE`

File: mono/metadata/handle.c [hunk 2]
```diff
@@ -17,204 +19,161 @@
...
 }
 
-MonoHandleArena**
-mono_handle_arena_current_addr (void)
+MonoArrayHandle
+mono_array_new_handle (MonoDomain *domain, MonoClass *eclass, uintptr_t n, MonoError *error)
 {
-	return (MonoHandleArena**) &mono_thread_info_current ()->handle_arena;
-}
+	return MONO_HANDLE_NEW (MonoArray, mono_array_new_checked (domain, eclass, n, error));
```
Comment (lambdageek) [inline mono/metadata/handle.c]: This is going to return a non-`mono_null_value_handle` even in the case of an error.

**Comment by kumpera** [comment]
@monojenkins merge

**Pull request merged**